x-pack/plugins/observability/server/services/slo/enrich_policy_refresh_task.ts

```diff
@@ -0,0 +1,118 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient, Logger, SavedObjectsClient } from '@kbn/core/server';
+import {
+  ConcreteTaskInstance,
+  TaskManagerSetupContract,
+  TaskManagerStartContract,
+} from '@kbn/task-manager-plugin/server';
+import { SLO_SUMMARY_ENRICH_POLICY_NAME } from '../../assets/constants';
+import { SO_SLO_TYPE } from '../../saved_objects';
+
+export const TASK_TYPE = 'SLO-Enrich-Policy-Refresh';
+
+export class EnrishPolicyRefreshTask {
+  private abortController = new AbortController();
+  private logger: Logger;
+  private taskManager?: TaskManagerStartContract;
+  private soClient?: SavedObjectsClient;
+  private esClient?: ElasticsearchClient;
+
+  constructor(taskManager: TaskManagerSetupContract, logger: Logger) {
+    this.logger = logger;
+
+    taskManager.registerTaskDefinitions({
+      [TASK_TYPE]: {
+        title: 'SLO Enrish Policy Refresh',
+        timeout: '1m',
+        maxAttempts: 1,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              const runAt = new Date().toISOString();
+              const lastRunAt = taskInstance.state?.lastRunAt || runAt;
+              if (!this.soClient) {
+                this.logger.debug('[SLO] soClient undefined, skipping task.');
+                return { state: { lastRunAt } };
+              }
+
+              if (!this.esClient) {
+                this.logger.debug('[SLO] esClient undefined, skipping task.');
+                return { state: { lastRunAt } };
+              }
+
+              const response = await this.soClient.find({
+                type: SO_SLO_TYPE,
+                filter: `slo.updated_at > "${lastRunAt}"`,
+                perPage: 1,
+              });
+              const shouldExecutePolicy = response.total > 0;
+
+              if (shouldExecutePolicy) {
+                this.logger.info(
+                  `[SLO] ${response.total} updated SLOs found since [${lastRunAt}]. Executing enrich policy.`
+                );
+
+                try {
+                  await this.esClient.enrich.executePolicy({
+                    name: SLO_SUMMARY_ENRICH_POLICY_NAME,
+                  });
+                } catch (err) {
+                  if (err?.meta?.body?.error?.type === 'es_rejected_execution_exception') {
+                    this.logger.info(
+                      `[SLO] Enrich policy [${SLO_SUMMARY_ENRICH_POLICY_NAME}] already executing.`
+                    );
+                    return { state: { lastRunAt } };
+                  }
+
+                  this.logger.info('[SLO] Cannot execute enrich policy. Will retry in next run.');
+                  return { state: { lastRunAt } };
+                }
+              }
+
+              return { state: { lastRunAt: runAt } };
```

Comment:
Does this task needs to run every 1 minute forever? from what i understand, we can run the task once kibana starts and trigger it again when any CRUD operations happens in routes. 

Otherwise task should go in stale mode.

Reply:
As for the task running every minute, I'm not too concerned about a task doing a very fast search and most of the time doing nothing else. Are you?

But I think If we schedule the task only after an operation on an SLO happened we could miss some. 
Let me explain: we create 2 SLO in series with a couple of seconds delay between them:
SLO 1 is created
SLO 1 triggers the task [task-1]
Task task-1 starts [execute policy is already executing]
SLO 2 is created [task-1 is still running]
SLO 2 cannot trigger task because task-1 is already in progress
Task task-1 ends
SLO 2 never triggers an execute policy

When an instance starts our SREs create hundreds of SLOs. This scenario could definitely happen.

Reply:
Makes sense, thanks for explanation.

Reply:
@kdelemme Shouldn't the execute policy eventually copy the updated value from saved object to the summary index? In following scenario I updated the name, but summary index didn't get updated with the new name. Saved object contains the updated name and updated names appears in the UI. Is there any special slo enrich policy index where the updated values are stored?

https://github.com/elastic/kibana/assets/2852703/088912a4-ab43-4f80-8e33-aadebf52d93b

Reply:
@kdelemme Ok interesting, something looks nasty. Let me describe you the scenario I tried and please let me know if you can reproduce as well.

The SLO I updated above, was a reset SLO, if that makes any difference. So I guess something is wrong not copying the updated SLO field (name in my case) to the summary index, if that was a reset SLO.

I thought let's create a branch new SLO and verify if the updated name will be copied to the `summary` index. Indeed it was copied in the summary index, but here's what else happened:
- the old (reset) SLO name was finally updated in the summary index.
- but guess what, 4 new SLOs appeared out of nowhere in the SLO list page. These 4 SLOs were previously created on main branch (`v2`), I didn't use the reset API, they simply appeared in the SLO list, once I created a brand new SLO

Can you try as well?

Reply:
I'll make a document with the scenarios I want to try:

- No previous data/running kibana in your branch
    - create an SLO
    - update the SLO
    - verify the summary index gets updated

- Previous data/running kibana in your branch
    - reset one SLO
    - update the SLO
    - verify that summary index gets updated for this SLO
    - verify that non reset SLOs don't get migrated (unless it is by design that they should be updated, I will not stop high cardinality this name and see if that makes a difference)


Let me know if the assumptions I make for what is expected are wrong.

Reply:
I finished the manual testing. Here's the [document](https://docs.google.com/document/d/1HHUrn_EV5g9CXxUYs_rl0NWXH14g9Ds3eo3z6uYBFRo/edit).

@kdelemme To summarize it, I would like a clarification regarding what I should expect for non reset SLOs and if they should automatically get re-indexed in the summary-v3 indices or not. From my testing it looks like they get indexed no matter if I reset them or not.

Reply:
I think this comment answers this question https://github.com/elastic/kibana/pull/171555#issuecomment-1825871059